src/backend/distributed/planner/shard_pruning.c

```diff
@@ -405,6 +523,150 @@ PruneShards(Oid relationId, Index rangeTableId, List *whereClauseList,
+		if (SAORestrictions(arrayOperatorExpression, partitionColumn, NULL))
+		{
+			return true;
+		}
+
+		return false;
+	}
+	else
+	{
+		return false;
+	}
+}
+
+
+/*
+ * Build a logical tree of valid constraints and invalid constaints for pruning.
+ */
+static bool
+BuildPruningTree(Node *node, PruningTreeBuildContext *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, List))
+	{
+		return expression_tree_walker(node, BuildPruningTree, context);
+	}
+	else if (IsA(node, BoolExpr))
+	{
+		BoolExpr *boolExpr = (BoolExpr *) node;
+		bool isAnded = boolExpr->boolop == AND_EXPR;
+
+		if (boolExpr->boolop == NOT_EXPR)
+		{
+			return false;
+		}
+		else if (context->current->isAnd != isAnded)
+		{
+			PruningTreeNode *child = CreatePruningNode(isAnded);
+
+			context->current->childBooleanNodes = lappend(
+				context->current->childBooleanNodes, child);
+
+			PruningTreeBuildContext newContext = { 0 };
+			newContext.partitionColumn = context->partitionColumn;
+			newContext.current = child;
+
+			return expression_tree_walker((Node *) boolExpr->args,
+										  BuildPruningTree, &newContext);
+		}
+		else
+		{
+			return expression_tree_walker(node, BuildPruningTree, context);
+		}
+	}
+	else if (IsValidConditionNode(node, context->partitionColumn))
+	{
+		context->current->validConstraints = lappend(context->current->validConstraints,
+													 node);
+
+		return false;
+	}
+	else
+	{
+		context->current->hasInvalidConstraints = true;
+
+		return false;
+	}
+}
+
+
+/*
+ * Simplifies the logical tree of valid and invalid constraints for pruning.
+ * The goal is to remove any node having just a single constraint associated with it.
+ * This constraint is assigned to the parent logical node.
+ * Removal of nodes is done by traversing from tree leafs upward.
+ *
+ * For example logical tree of
+ * AND(hash_col = 1, OR(X)) gets simplified into AND(hash_col = 1, X)
+ * Where X is any unknown condition.
+ */
+static void
+SimplifyPruningTree(PruningTreeNode *node, PruningTreeNode *parent)
+{
+	/* Copy list of children as its mutated inside the loop */
+	List *childBooleanNodes = list_copy(node->childBooleanNodes);
+
+	ListCell *cell;
+	foreach(cell, childBooleanNodes)
+	{
+		PruningTreeNode *child = (PruningTreeNode *) lfirst(cell);
+		SimplifyPruningTree(child, node);
+	}
+
+	if (!parent)
+	{
+		/* Root is always ANDed expressions */
+		Assert(IsAndOp(node));
+		return;
+	}
+
+	/* Boolean operator with just a single (regocnized/unknown) constraints gets simplified */
+	if (ConstraintCount(node) <= 1)
+	{
+		parent->validConstraints = list_concat(parent->validConstraints,
+											   node->validConstraints);
+		parent->hasInvalidConstraints = parent->hasInvalidConstraints ||
+										node->hasInvalidConstraints;
+
+		/* Remove current node from parent. Its constraint was assigned to the parent above */
+		parent->childBooleanNodes = list_delete_ptr(parent->childBooleanNodes, node);
```

Comment:
Could instead copy into a new filtered list instead of deleting from the old one. Might be a bit more optimal. This just seemed easier to eyes like this